### PR TITLE
feat: remove @types/eslint package

### PR DESCRIPTION
**Pull request by snitin315**
Reference https://github.com/eslint/eslint/issues/18928

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If removing a declaration:

- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: notNeededPackages.json [hunk 1]
```diff
@@ -1641,6 +1641,10 @@
             "libraryName": "escape-string-regexp",
             "asOfVersion": "2.0.0"
         },
+        "eslint": {
+            "libraryName": "eslint",
+            "asOfVersion": "9.11.1"
```
Comment (snitin315) [inline notNeededPackages.json:1646]: Actually, types were added in [v9.10.0,](https://github.com/eslint/eslint/releases/tag/v9.10.0), But I've used [v9.11.1](https://github.com/eslint/eslint/releases/tag/v9.11.1) as includes a patch fix (https://github.com/eslint/eslint/pull/18938) for the types. 

Or should I use v9.10.0 only?

// @JoshuaKGoldberg @nzakas

Comment (JoshuaKGoldberg) [reply notNeededPackages.json:1646]: I'd go with `v9.10.0`. That's when they _existed_. 18938 fixes a bug that impacts some users, but that's a _bug_ and not an _absence_.

**Review by JoshuaKGoldberg — APPROVED** [review]
🔪

File: notNeededPackages.json (hunk 1 above)
Comment (snitin315) [reply notNeededPackages.json:1646]:
```suggestion
            "asOfVersion": "9.10.0"
```

**Review by JoshuaKGoldberg — APPROVED** [review]
🔪🔪

**Review by bradzacher — APPROVED** [review]

**Comment by snitin315** [comment]
@sandersn Done.

**Review by sandersn — APPROVED** [review]

**Comment by karlhorky** [comment]
Nice, thanks for this @snitin315 and others involved!

What happens with [the existing `@types/eslint` package on npm](https://www.npmjs.com/package/@types/eslint) now?

It is currently still published at version `9.6.1` (last publish: 26 Aug 2024) without any warnings or deprecation notices:

Screenshot 2024-10-17 at 12 05 19

Or maybe there are plans already which I missed to manually deprecate this from npm...?

**Comment by pauliesnug** [comment]
@karlhorky I believe usually after some time the DT bot deprecates the package and adds the warning `This is a stub types definition. eslint provides its own type definitions, so you do not need this installed.`. Maybe this was granted a special exception as it is such a popular types package?

**Comment by sandersn** [comment]
That's right. My explanation is at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70735#issuecomment-2392047959

**Comment by karlhorky** [comment]
@sandersn oh, yeah I saw that comment.

Does that mean that the obsolete `@types/eslint` package will stay online without any notice to users?

If yes, seems like this will lead to confusion for new projects and existing projects, and unnecessary dependencies being installed...?

**Comment by pauliesnug** [comment]
@karlhorky Yeah I agree, I think instead of not publishing a stub and not publishing a deprecation notice, we should _only_ skip the stub, but keep the notice

**Comment by sandersn** [comment]
@karlhorky `@types/eslint` stays online with the last version at 9.6.1 -- no more publishes.

@pauliesnug I like the idea of a custom deprecate notice, but there's no place to put it unless we publish another version of `@types/eslint` -- `eslint@9.6.0` doesn't ship types, so it makes sense for people to install `@types/eslint@9.6`. I guess we could re-publish 9.6.1 manually as 9.10.0, but that seems misleading as well.

**Comment by karlhorky** [comment]
> I like the idea of a custom deprecate notice, but there's no place to put it unless we publish another version of `@types/eslint`

What about [`npm deprecate`](https://docs.npmjs.com/cli/v10/commands/npm-deprecate)? That wouldn't require publishing another version I guess, would it?

**Comment by sandersn** [comment]
What I mean is, `npm deprecate @types/eslint@9.6.1` doesn't make sense, because `eslint@9.6` doesn't ship types. So `@types/eslint@9.6.1` shouldn't be deprecated. It still makes sense to install.

**Comment by karlhorky** [comment]
Ok, I think I'm understanding what you mean.

To be clear, what I think could be a common user flow:

1. User is setting up a new project with ESLint, and relies on their memory
2. User remembers that for ESLint, the types are not included and `@types/eslint` is required
3. User checks npm page `@types/eslint` to be sure (does not look at version number)
4. User does not see any deprecation notice or anything that says that `@types/eslint` is obsolete now (and also doesn't go looking through old closed issues / PRs to investigate)
5. 💥 User installs `@types/eslint` along with ESLint - an obsolete, unnecessary package

**Comment by sandersn** [comment]
The bad thing that happens at (5) is not a crash, it's an extra 253 KB in node_module that the compiler ignores, plus an unnecessary entry in package.json. It is bad that the JS ecosystem continues to accrue cruft.

Also, deprecating stub packages got even less useful recently with https://github.com/npm/npm-pick-manifest/pull/33; npm now never resolves to a deprecated package when any un-deprecated one exists. So `npm i @types/eslint` would always give 9.6, not the 9.10 stub. Maybe what we could do is publish a final patch release, 9.6.2, that's a copy of 9.6.1 with an updated README that adds "if you're using a version of eslint  9.10 or newer, you don't need these types." That would stop your scenario at (3).

However, I don't think most people will `npm i @types/eslint@9.10` or check the types README, so neither will reach the majority of people.

**Comment by kachkaev** [comment]
How about releasing a stub `@types/eslint@10.0.0` package? 9.6.1 (or 9.6.2) will be the last un-deprecated package with actual typings in place.

**Comment by sandersn** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70735#issuecomment-2392047959. Briefly, types packages depend on each other with `*`, so they'll get the stub instead of working types -- at least on old versions of npm. Newer versions of npm will resolve to a working version. But...so will people when they `npm i @types/eslint`.

**Comment by jakebailey** [comment]
The flip side is that without the stub, all packages on DT will only ever test the DT package, because they'll never know to download `eslint` instead. That's basically what happened when this PR was merged, so I'm not sure _that's_ a good result either... This hole didn't exist before we switched to a real pnpm monorepo, because DT would not have been able to pull older types from npm without interaction. Things could be very broken right now with people using `eslint` directly and we'd have no idea.

**Comment by jakebailey** [comment]
I recently merged in support for peer deps; given eslint packages are extensions of eslint and therefore would be peer deps, perhaps we should fix all of those to be peer deps on `@types/eslint` and then publish the stub? That way, a user can fairly easily install the "right" types?

(But, doing so will forcibly install `@types/eslint` even when it's not needed due to it being eslint v9, so, it's all terrible. That's not different than today, though.)

**Pull request merged**